File: src/org/thoughtcrime/securesms/logging/PersistentLogger.java
package org.thoughtcrime.securesms.logging;

import android.content.Context;
import android.support.annotation.AnyThread;
import android.support.annotation.WorkerThread;

import org.thoughtcrime.securesms.database.NoExternalStorageException;
import org.thoughtcrime.securesms.util.concurrent.ListenableFuture;
import org.thoughtcrime.securesms.util.concurrent.SettableFuture;

import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.IOException;
import java.io.PrintStream;
import java.text.SimpleDateFormat;
import java.util.Arrays;
import java.util.Date;
import java.util.LinkedList;
import java.util.List;
import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class PersistentLogger extends Log.Logger {

  private static final String TAG     = PersistentLogger.class.getSimpleName();

  private static final String LOG_V   = "V";
  private static final String LOG_D   = "D";
  private static final String LOG_I   = "I";
  private static final String LOG_W   = "W";
  private static final String LOG_E   = "E";
  private static final String LOG_WTF = "A";

  private static final String           LOG_DIRECTORY   = "log";
  private static final String           FILENAME_PREFIX = "log-";
  private static final int              MAX_LOG_FILES   = 5;
  private static final int              MAX_LOG_SIZE    = 300 * 1024;
  private static final SimpleDateFormat DATE_FORMAT     = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS zzz");

  private final Context  context;
  private final Executor executor;
  private final byte[]   secret;

  private LogFile.Writer writer;

  public PersistentLogger(Context context) {
    this.context  = context.getApplicationContext();
    this.secret   = LogSecretProvider.getOrCreateAttachmentSecret(context);
    this.executor = Executors.newSingleThreadExecutor(r -> {
      Thread thread = new Thread(r, "logger");
      thread.setPriority(Thread.MIN_PRIORITY);
      return thread;
    });

    executor.execute(this::initializeWriter);
  }

  @Override
  public void v(String tag, String message, Throwable t) {
    write(LOG_V, tag, message, t);
  }

  @Override
  public void d(String tag, String message, Throwable t) {
    write(LOG_D, tag, message, t);
  }

  @Override
  public void i(String tag, String message, Throwable t) {
    write(LOG_I, tag, message, t);
  }

  @Override
  public void w(String tag, String message, Throwable t) {
    write(LOG_W, tag, message, t);
  }

  @Override
  public void e(String tag, String message, Throwable t) {
    write(LOG_E, tag, message, t);
  }

  @Override
  public void wtf(String tag, String message, Throwable t) {
    write(LOG_WTF, tag, message, t);
  }

  @WorkerThread
  public ListenableFuture<String> getLogs() {
    final SettableFuture<String> future = new SettableFuture<>();

    executor.execute(() -> {
      StringBuilder builder = new StringBuilder();

      try {
        File[] logs = getSortedLogFiles();
        for (int i = logs.length - 1; i >= 0; i--) {
          try {
            LogFile.Reader reader = new LogFile.Reader(secret, logs[i]);
            builder.append(reader.readAll());
          } catch (IOException e) {
            android.util.Log.w(TAG, "Failed to read log at index " + i + ". Removing reference.");
            logs[i].delete();
          }
        }

        future.set(builder.toString());
      } catch (NoExternalStorageException e) {
        future.setException(e);
      }
    });

    return future;
  }

  @WorkerThread
  private void initializeWriter() {
    try {
      writer = new LogFile.Writer(secret, getOrCreateActiveLogFile());
    } catch (NoExternalStorageException | IOException e) {
      android.util.Log.e(TAG, "Failed to initialize writer.", e);
    }
  }

  @AnyThread
  private void write(String level, String tag, String message, Throwable t) {
    executor.execute(() -> {
      try {
        if (writer == null) {
          return;
        }

        if (writer.getLogSize() >= MAX_LOG_SIZE) {
          writer.close();
          writer = new LogFile.Writer(secret, createNewLogFile());
          trimLogFilesOverMax();
        }

        for (String entry : buildLogEntries(level, tag, message, t)) {
          writer.writeEntry(entry);
        }

      } catch (NoExternalStorageException e) {
        android.util.Log.w(TAG, "Cannot persist logs.", e);
      } catch (IOException e) {
        android.util.Log.w(TAG, "Failed to write line. Deleting all logs and starting over.");
        deleteAllLogs();
        initializeWriter();
      }
    });
  }

  private void trimLogFilesOverMax() throws NoExternalStorageException {
    File[] logs = getSortedLogFiles();
    if (logs.length > MAX_LOG_FILES) {
      for (int i = MAX_LOG_FILES; i < logs.length; i++) {
        logs[i].delete();
      }
    }
  }

  private void deleteAllLogs() {
    try {
      File[] logs = getSortedLogFiles();
      for (File log : logs) {
        log.delete();
      }
    } catch (NoExternalStorageException e) {
      android.util.Log.w(TAG, "Was unable to delete logs.", e);
    }
  }

  private File getOrCreateActiveLogFile() throws NoExternalStorageException {
    File[] logs = getSortedLogFiles();
    if (logs.length > 0) {
      return logs[0];
    }

    return createNewLogFile();
  }

  private File createNewLogFile() throws NoExternalStorageException {
    return new File(getOrCreateLogDirectory(), FILENAME_PREFIX + System.currentTimeMillis());
  }

  private File[] getSortedLogFiles() throws NoExternalStorageException {
    File[] logs = getOrCreateLogDirectory().listFiles();
    if (logs != null) {
      Arrays.sort(logs, (o1, o2) -> o2.getName().compareTo(o1.getName()));
      return logs;
    }
    return new File[0];
  }

  private File getOrCreateLogDirectory() throws NoExternalStorageException {
    File logDir = new File(context.getCacheDir(), LOG_DIRECTORY);
    if (!logDir.exists() && !logDir.mkdir()) {
      throw new NoExternalStorageException("Unable to create log directory.");
    }

    return logDir;
  }

  private List<String> buildLogEntries(String level, String tag, String message, Throwable t) {
    List<String> entries = new LinkedList<>();
    Date         date    = new Date();

    entries.add(buildEntry(level, tag, message, date));

    if (t != null) {
      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
      t.printStackTrace(new PrintStream(outputStream));

      String   trace = new String(outputStream.toByteArray());
      String[] lines = trace.split("\\n");

      for (String line : lines) {
        entries.add(buildEntry(level, tag, line, date));
      }
    }

    return entries;
  }

  private String buildEntry(String level, String tag, String message, Date date) {
    return DATE_FORMAT.format(date) + ' ' + level + ' ' + tag + ": " + message;
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/logging/PersistentLogger.kt
package org.thoughtcrime.securesms.logging

import android.app.Application
import android.os.Looper
import org.signal.core.util.logging.Log
import org.signal.core.util.logging.Scrubber
import org.thoughtcrime.securesms.BuildConfig
import org.thoughtcrime.securesms.database.LogDatabase
import org.thoughtcrime.securesms.database.model.LogEntry
import org.thoughtcrime.securesms.logging.PersistentLogger.LogRequest
import org.thoughtcrime.securesms.logging.PersistentLogger.WriteThread
import java.io.ByteArrayOutputStream
import java.io.PrintStream
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/**
 * A logger that will persist log entries in [LogDatabase].
 *
 * We log everywhere, and we never want it to slow down the app, so performance is critical here.
 * This class takes special care to do as little as possible on the main thread, instead letting the background thread do the work.
 *
 * The process looks something like:
 * - Main thread creates a [LogRequest] object and puts it in a queue
 * - The [WriteThread] constantly pulls from that queue, formats the logs, and writes them to the database.
 */
class PersistentLogger(
  application: Application
) : Log.Logger() {

  companion object {
    private const val LOG_V = "V"
    private const val LOG_D = "D"
    private const val LOG_I = "I"
    private const val LOG_W = "W"
    private const val LOG_E = "E"
  }

  private val logEntries = LogRequests()
  private val logDatabase = LogDatabase.getInstance(application)
  private val cachedThreadString: ThreadLocal<String> = ThreadLocal()

  init {
    WriteThread(logEntries, logDatabase).apply {
      priority = Thread.MIN_PRIORITY
    }.start()
  }

  override fun v(tag: String?, message: String?, t: Throwable?, keepLonger: Boolean) {
    write(LOG_V, tag, message, t, keepLonger)
  }

  override fun d(tag: String?, message: String?, t: Throwable?, keepLonger: Boolean) {
    write(LOG_D, tag, message, t, keepLonger)
  }

  override fun i(tag: String?, message: String?, t: Throwable?, keepLonger: Boolean) {
    write(LOG_I, tag, message, t, keepLonger)
  }

  override fun w(tag: String?, message: String?, t: Throwable?, keepLonger: Boolean) {
    write(LOG_W, tag, message, t, keepLonger)
  }

  override fun e(tag: String?, message: String?, t: Throwable?, keepLonger: Boolean) {
    write(LOG_E, tag, message, t, keepLonger)
  }

  override fun flush() {
    logEntries.blockForFlushed()
  }

  private fun write(level: String, tag: String?, message: String?, t: Throwable?, keepLonger: Boolean) {
    logEntries.add(LogRequest(level, tag ?: "null", message, System.currentTimeMillis(), getThreadString(), t, keepLonger))
  }

  private fun getThreadString(): String {
    var threadString = cachedThreadString.get()

    if (cachedThreadString.get() == null) {
      threadString = if (Looper.myLooper() == Looper.getMainLooper()) {
        "main "
      } else {
        String.format("%-5s", Thread.currentThread().id)
      }

      cachedThreadString.set(threadString)
    }

    return threadString!!
  }

  private data class LogRequest(
    val level: String,
    val tag: String,
    val message: String?,
    val createTime: Long,
    val threadString: String,
    val throwable: Throwable?,
    val keepLonger: Boolean
  )

  private class WriteThread(
    private val requests: LogRequests,
    private val db: LogDatabase
  ) : Thread("signal-logger") {

    private val dateFormat = SimpleDateFormat("yyyy-MM-dd HH:mm:ss.SSS zzz", Locale.US)
    private val buffer = mutableListOf<LogRequest>()

    override fun run() {
      while (true) {
        requests.blockForRequests(buffer)
        db.insert(buffer.flatMap { requestToEntries(it) }, System.currentTimeMillis())
        buffer.clear()
        requests.notifyFlushed()
      }
    }

    fun requestToEntries(request: LogRequest): List<LogEntry> {
      val out = mutableListOf<LogEntry>()

      val createDate = Date(request.createTime)

      out.add(
        LogEntry(
          createdAt = request.createTime,
          keepLonger = request.keepLonger,
          body = formatBody(request.threadString, createDate, request.level, request.tag, request.message)
        )
      )

      if (request.throwable != null) {
        val outputStream = ByteArrayOutputStream()
        request.throwable.printStackTrace(PrintStream(outputStream))

        val trace = String(outputStream.toByteArray())
        val lines = trace.split("\\n".toRegex()).toTypedArray()

        val entries = lines.map { line ->
          LogEntry(
            createdAt = request.createTime,
            keepLonger = request.keepLonger,
            body = formatBody(request.threadString, createDate, request.level, request.tag, line)
          )
        }

        out.addAll(entries)
      }

      return out
    }

    fun formatBody(threadString: String, date: Date, level: String, tag: String, message: String?): String {
      return "[${BuildConfig.VERSION_NAME}] [$threadString] ${dateFormat.format(date)} $level $tag: ${Scrubber.scrub(message ?: "")}"
    }
  }

  private class LogRequests {
    val logs = mutableListOf<LogRequest>()
    val logLock = Object()

    var flushed = false
    val flushedLock = Object()

    fun add(entry: LogRequest) {
      synchronized(logLock) {
        logs.add(entry)
        logLock.notify()
      }
    }

    /**
     * Blocks until requests are available. When they are, the [buffer] will be populated with all pending requests.
     * Note: This method gets hit a *lot*, which is why we're using a buffer instead of spamming out new lists every time.
     */
    fun blockForRequests(buffer: MutableList<LogRequest>) {
      synchronized(logLock) {
        while (logs.isEmpty()) {
          logLock.wait()
        }

        buffer.addAll(logs)
        logs.clear()
        flushed = false
      }
    }

    fun blockForFlushed() {
      synchronized(flushedLock) {
        while (!flushed) {
          flushedLock.wait()
        }
      }
    }

    fun notifyFlushed() {
      synchronized(flushedLock) {
        flushed = true
        flushedLock.notify()
      }
    }
  }
}
